src/Orleans/Streams/PubSub/MemoryPubSubStore.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Streams.PubSub
+{
+    [Serializable]
+    internal class MemoryPubSubStreamSubscriptions
+    {
+        // use ConcurrentDictionary as a concurrent HasSet
+        //http://stackoverflow.com/questions/18922985/concurrent-hashsett-in-net-framework
+        private readonly ConcurrentDictionary<PubSubPublisherState, byte> producers;
+        private readonly ConcurrentDictionary<PubSubSubscriptionState, byte> consumers;
+
+        public MemoryPubSubStreamSubscriptions()
+        {
+            producers = new ConcurrentDictionary<PubSubPublisherState, byte>();
+            consumers = new ConcurrentDictionary<PubSubSubscriptionState, byte>();
+        }
+
+        public void AddProducer(PubSubPublisherState state)
+        {
+            producers.AddOrUpdate(state, byte.MinValue, (publisherState, b) => b);
+        }
+
+        public void RemoveProducer(PubSubPublisherState state)
+        {
+            byte b;
+            bool re = producers.TryRemove(state, out b);
+            if (!re)
+            {
+                throw new OrleansException($"{this.GetType().Name} : Remove producer failed");
+            }
+        }
+
+        public void AddConsumer(PubSubSubscriptionState state)
+        {
+            consumers.AddOrUpdate(state, byte.MinValue, (subscriptionState, b) => b);
+        }
+
+        public void RemoveConsumer(PubSubSubscriptionState state)
+        {
+            byte b;
+            bool re = consumers.TryRemove(state, out b);
+            if (!re)
+            {
+                throw new OrleansException($"{this.GetType().Name} : Remove consumer failed");
+            }
+        }
+
+        public ICollection<PubSubPublisherState> Producers => producers.Keys;
+        public ICollection<PubSubSubscriptionState> Consumers => consumers.Keys;
+    }
+    
+    internal class MemoryPubSubStore
```

Comment:
Unless I'm mistaken, this is mostly the logic from the rendezvous grain, correct?  Can this logic be shared?